### PR TITLE
Merge in project-fuzzball

### PR DESCRIPTION
**Pull request by rtfeldman**
See https://github.com/project-fuzzball/test/issues/5 for details!

TODO before merging:

* Give folks some time to review this
* Revise README (it can link to [legacy-elm-test](https://github.com/rtfeldman/legacy-elm-test), [html-test-runner](https://github.com/rtfeldman/html-test-runner), and [node-test-runner](https://github.com/rtfeldman/node-test-runner) even though they won't be published until after this has been published and they can depend on it.)

Note: Since this was a scratch-rewrite, I changed the license to BSD.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,158 +1,3 @@
...
-language: haskell
-install:
-  - npm install -g elm
-  - elm-package install -y
-before_script: 
-  - elm-make --yes --output test.js tests/Tests.elm
-script: node test.js
-```
-You can look at the `.travis.yml` file in this repository to see a real example.
+Write tests in Elm.
```
Comment (tianyu) [inline README.md]: This still needs some expanding

Comment (rtfeldman) [inline README.md]: Agreed! It's on my TODO list before merging this. 😄

File: LICENSE [hunk 2]
```diff
@@ -1,20 +1,28 @@
...
+* Redistributions of source code must retain the above copyright notice, this
+  list of conditions and the following disclaimer.
 
-The above copyright notice and this permission notice shall be included in all
-copies or substantial portions of the Software.
+* Redistributions in binary form must reproduce the above copyright notice,
+  this list of conditions and the following disclaimer in the documentation
+  and/or other materials provided with the distribution.
+
+* Neither the name of elm-stylesheets nor the names of its
```
Comment (gampleman) [inline LICENSE]: This is also the wrong project name.

File: src/Expect.elm [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+fail : String -> Expectation
+fail =
+    Test.Expectation.Fail ""
+
+
+{-| Return `Nothing` if the given [`Expectation`](#Expectation) is a [`pass`](#pass).
+
+If it is a [`fail`](#fail), return a record containing the failure message,
+along with the given inputs if it was a fuzz test. (If no inputs were involved,
+the record's `given` field will be `""`).
```
Comment (gampleman) [inline src/Expect.elm]: Wouldn't a Maybe make more sense as a type for the given?

```
getFailure : Expectation -> Maybe { given : Maybe String, message : String }
```

File: LICENSE (hunk 2 above)
Comment (rtfeldman) [inline LICENSE]: Thanks! Fixed in https://github.com/elm-community/elm-test/pull/24/commits/f72463fc647bf277c05fa38786c6c8a59dc2a28b

File: src/Expect.elm (hunk 3 above)
Comment (rtfeldman) [inline src/Expect.elm]: It would, but the overhead associated with storing that for every single test didn't seem worth the slight semantic distinction.

Currently there's no way to get a `""` value for `given` other than running a non-fuzz test, so I thought about translating that to `Nothing` and anything else to `Just`, but then it seemed like I'd be setting up a false expectation if it ended up being possible in the future for `given` to be `""` for a fuzz test.

tl;dr I agree, but the juice doesn't seem worth the squeeze. 😄

File: src/Fuzz.elm [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+{-| A fuzzer for int values within between a given minimum and maximum value,
+inclusive. Shrunken values will also be within the range.
+-}
+intRange : Int -> Int -> Fuzzer Int
+intRange min max =
+    custom (Random.int min max)
+        (Shrink.keepIf (\i -> i >= min && i <= max) Shrink.int)
+
+
+{-| A fuzzer for float values. It will never fuzzuce `NaN`, `Infinity`, or `-Infinity`.
```
Comment (gampleman) [inline src/Fuzz.elm]: fuzzuce? That's quite the neologism.

File: src/Fuzz.elm [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+
+{-| Turn a 5-tuple of fuzzers into a fuzzer of 5-tuples.
+-}
+tuple5 : ( Fuzzer a, Fuzzer b, Fuzzer c, Fuzzer d, Fuzzer e ) -> Fuzzer ( a, b, c, d, e )
+tuple5 ( Internal.Fuzzer fuzzA, Internal.Fuzzer fuzzB, Internal.Fuzzer fuzzC, Internal.Fuzzer fuzzD, Internal.Fuzzer fuzzE ) =
+    custom (Random.map5 (,,,,) fuzzA.generator fuzzB.generator fuzzC.generator fuzzD.generator fuzzE.generator)
+        (Shrink.tuple5 ( fuzzA.shrinker, fuzzB.shrinker, fuzzC.shrinker, fuzzD.shrinker, fuzzE.shrinker ))
+
+
+{-| Map a function over a fuzzer. This works exactly like `convert`,
```
Comment (gampleman) [inline src/Fuzz.elm]: `convert` is mentioned, but not exposed in the module.

File: src/Expect.elm (hunk 3 above)
Comment (gampleman) [inline src/Expect.elm]: Fair enough.

File: src/Fuzz.elm [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+            shrinker question =
+                case question of
+                    Name n ->
+                        Shrink.string n |> Shrink.map Name
+
+                    Age i ->
+                        Shrink.int i |> Shrink.map Age
+        in
+            Fuzz.custom generator shrinker
+-}
```
Comment (evancz) [inline src/Fuzz.elm]: I think this big long example goes with `custom`, not with the type.

I like to structure docs so that the relevant type comes first with a simple intuitive explanation. Details come out later in subsequent documentation. See [Maybe](http://package.elm-lang.org/packages/elm-lang/core/4.0.1/Maybe) or `Body` in [Http](http://package.elm-lang.org/packages/evancz/elm-http/3.0.1/Http).

File: src/Test.elm [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+See [`test`](#test) and [`fuzz`](#fuzz) for some ways to create a `Test`.
+-}
+type alias Test =
+    Internal.Test
+
+
+{-| Run the given tests in order.
+
+    concat [ testDecoder, testSorting ]
+-}
```
Comment (evancz) [inline src/Test.elm]: Is it in-order if you can run multiple threads?

File: src/Test.elm [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+concat : List Test -> Test
+concat =
+    Internal.Batch
+
+
+{-| Remove any test unless its description satisfies the given predicate
+function. Nested descriptions added with [`describe`](#describe) are not considered.
+-}
```
Comment (evancz) [inline src/Test.elm]: The second sentence is trying to clarify which descriptions are considered, but I don't get it. Only the leafs? Also internal nodes? Which ones *are* considered?

File: src/Test.elm [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+
+-- INTERNAL HELPERS --
+
+
+defaults : { runs : Int, seed : Random.Seed }
+defaults =
+    { runs = 100
+    , seed = Random.initialSeed 42
+    }
```
Comment (evancz) [inline src/Test.elm]: Appears to be unused.

File: src/Expect.elm [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+        |> Expect.equal 5
+
+
+    {-
+
+    0
+    ╷
+    │ Expect.equal
+    ╵
+    5
```
Comment (evancz) [inline src/Expect.elm]: Maybe use an example with larger bad values? Like two strings that are not equal. This way, it is clearer why the layout is like this. As is, it's like "why not just put them side-by-side?"

File: src/Test.elm (hunk 7 above)
Comment (rtfeldman) [inline src/Test.elm]: Yeah, I was debating whether this would be worth maintaining, and decided it would be for a consistent UX.

When I see a different failed test at the top of my failures list, I think "cool, fixed it, time to context-switch to the next failure." If it turned out this was just that things had finished in a different order, it sounds like it would lead to frustration pretty easily.

From an implementation standpoint this means we can't always report failures quite as soon as we know about them. I doubt the delays will be noticeable in practice, whereas I think inconsistent ordering would get annoying fast.

File: src/Test.elm (hunk 8 above)
Comment (rtfeldman) [inline src/Test.elm]: Yeah I think an example is necessary...thanks for reminding me though - I'd forgotten about this TODO. 😄

File: src/Fuzz.elm (hunk 5 above)
Comment (rtfeldman) [inline src/Fuzz.elm]: Good point! Thanks for surfacing. 👍

File: src/Fuzz.elm (hunk 4 above)
Comment (rtfeldman) [inline src/Fuzz.elm]: 😅  these used to be called "Producers" and find/replace happened and...yeah.

File: src/Fuzz.elm (hunk 6 above)
Comment (rtfeldman) [inline src/Fuzz.elm]: Seems reasonable!

File: src/Test/Runner/Log.elm [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+import Random
+import Test exposing (Test)
+import Test.Runner.String exposing (Summary)
+import String
+
+
+{-| Run the test using the default `Test.Runner.String` options.
+-}
+run : Test -> a -> a
```
Comment (evancz) [inline src/Test/Runner/Log.elm]: I would expect this to be `Test -> Task x ()`. Right now folks are using `Debug.log` to hack around the fact that there is not a proper console library yet. There will be, probably with cross-platform colors and all that too. So I'd say keep the implementation as is, but wrap it up as a task such that it runs when the task is run.

Based on that, I think [this code](https://github.com/project-fuzzball/test/blob/e294e6a62099eca1a3caf34f70b3a503904ca750/examples/LogRunnerExample.elm#L21-L28) should be pretty different. With a threaded implementation, things may be "out working" for quite some time (whoo tasks!) so you would want to hear about the results only when it's done, and display something else in the meantime. So I'd expect this task to be run in `init`. Once I am done with the server-side rendering stuff, there should be a function in `core` that produces `Program` values without an `html` dependency, so I'd plan for that reality, even if it means having a `elm-lang/virtual-dom` dependency temporarily.

**Comment by evancz** [comment]
This is looking great!

I feel very strongly about https://github.com/elm-community/elm-test/pull/24/files#r70340846, but otherwise, things are looking really nice :)

**Aside:** When you demo'd fuzz testing to me, I saw an error where the args were in a tuple. I didn't get why it was printed that way. I figured it was just weird. Now that I have seen the implementation of `fuzz2` I get it. I know there is a way to handle your `fuzzN` functions specially to get it to print out nicer, and maybe that'd be worthwhile at some later time. Maybe never. Unclear. Def not a blocker!

**Comment by rtfeldman** [comment]
I thought about making fuzzN do something special, but then I thought about
how it would mean custom stuff where fuzzN wasn't the best choice would
always have different-looking output. That would incentivize doing
contortions to use fuzzN no matter what, even if using tuples made more
sense from a code perspective.

I figured if all fuzzN did was save you from making a tuple and
destructuring it, the story around "roll your own" is just that it's more
effort, not that it's a strange-looking result.

File: src/Test/Runner/Log.elm (hunk 11 above)
Comment (rtfeldman) [inline src/Test/Runner/Log.elm]: Seems reasonable! The idea behind this was that this repo should offer some way for you to try out running tests without needing a full-fledged HTML runner or Node runner.

However, I think realistically we can get by having the example do `Debug.log` on its own (in a browser, since we can already run Tasks there), and then let people use the String runner from `elm-repl` if they want to try it out on a CLI.

File: src/Test.elm (hunk 8 above)
Comment (gampleman) [inline src/Test.elm]: Might help clarify what the use case for this is.

File: src/Test/Runner/Log.elm (hunk 11 above)
Comment (evancz) [inline src/Test/Runner/Log.elm]: Ah, I think I get the intent then. Do I understand correctly that the proposed change is to let people call `Debug.log` themselves rather than encoding it into the API with a helper function?

File: src/Test.elm (hunk 7 above)
Comment (evancz) [inline src/Test.elm]: I agree with all that. Here is my understanding: the results would be displayed in-order, even though the evaluation may be interleaved. If true, that seems ideal / I feel like pointing out that they run in-order draws attention to something everyone would *expect* is the default anyway, making it feel weird to mention.

Comment (rtfeldman) [inline src/Test.elm]: Fair point! I'll remove that sentence.

File: src/Test/Runner/Log.elm (hunk 11 above)
Comment (rtfeldman) [inline src/Test/Runner/Log.elm]: I'd like for there to be a default, but I think it's important that they live in different packages so they can be versioned differently.

For example, if `elm-test` depends on `html`, and my app does too, it's possible to find myself in a situation where I can no longer use the latest `elm-test` because I have not yet upgraded my application to use the latest `html`...even though I am not using the HTML test runner.

(We ran into this back in like the 0.15-0.16 days. It was a pain.)

My intention was to provide (and maintain myself) the default implementations and link to them from the README. Maybe I could use them for the examples too, come to think of it...at which point there's no need for the `String` runner.

File: src/Fuzz.elm (hunk 4 above)
Comment (rtfeldman) [inline src/Fuzz.elm]: Fixed in 110c138053e9274b2dde22ff5c3f28ed708b8a6f

File: src/Fuzz.elm (hunk 5 above)
Comment (rtfeldman) [inline src/Fuzz.elm]: Fixed in 6288a938680d926056c4d53185b477abecf29e96

File: src/Test.elm (hunk 7 above)
Comment (rtfeldman) [inline src/Test.elm]: Changed in 3529b3747097f29d74135595b048ad79139c743b

File: src/Test/Runner/Log.elm (hunk 11 above)
Comment (rtfeldman) [inline src/Test/Runner/Log.elm]: I ended up just moving the Log and String runners into `examples`. This way people can still try quickly things out with `LogRunnerExample` if they like.

I didn't bother updating the API to use a `Task`, since it's no longer publicly exposed.

**Comment by mgold** [comment]
Evan's link that he feels strongly about no longer works; I take it that the concern was addressed or respected?

**Comment by rtfeldman** [comment]
Yeah, I went through and addressed all the comments. 😄
